[shopsys] Elastic search filtering (#943)
shopsys/shopsys: docs/model/product-search-via-elasticsearch.md

The directory is configured using `%shopsys.elasticsearch.structure_dir%` parameter.

### Product data export
No data are automatically stored in Elasticsearch by "itself". When you store data into a relational database, they are not stored in Elasticsearch.
You have to export data from database into Elasticsearch actively.

> Following product attributes are exported into Elasticsearch (i.e. the search or filtering is performed on these fields only):

Reviewer: ```suggestion
Following product attributes are exported into Elasticsearch (i.e. the search or filtering can be performed on these fields only):
```

* name
* catnum
* partno
* ean
* description
* short description
* flags
* brand
* categories

> * prices

Reviewer: what prices? Please be more descriptive because pricing is complex problem

> * in_stock

Reviewer: Number in stock or availability?

Reviewer: CAlculated availability?

Author: I described it more

* parameters
* ordering_priority

> * calculated_selling_denied

Reviewer: And calculated displayability (or how it is called)?

Author: It's not needed as not visible products are not exported to elasticsearch

Aggregation numbers are counted with help of Elasticsearch too thanks to methods `Shopsys\FrameworkBundle\Model\Product\ProductOnCurrentDomainElasticFacade::getProductFilterCountDataInCategory` and
`Shopsys\FrameworkBundle\Model\Product\ProductOnCurrentDomainElasticFacade::getProductFilterCountDataForSearch`.

List of choices (exact parameters, brands, flags) is loaded from PostgreSQL as there is no benefit from loading them from Elasticsearch.

## Implementation

`ProductController` and `SearchController` uses `Shopsys\FrameworkBundle\Model\Product\ProductOnCurrentDomainFacadeInterface` to get filtered products.

Currently there are two implementations of `Shopsys\FrameworkBundle\Model\Product\ProductOnCurrentDomainFacadeInterface`:
 - `ProductOnCurrentDomainFacade`
    - filters data through SQL
    - slower than Elasticsearch, but on the other hand can be used easily on more complex pricing models (for example exact price is calculated with SQL function)
 - `ProductOnCurrentDomainElasticFacade`
    - filters data through Elasticsearch
    - much faster than filtering through SQL and remain fast independently of the number of selected filters

> Class used can be easily changed in `services.yml` file in your project.

Reviewer 2: Do you think this is enough or should we have information about what you need to change in the `services.yml` file to change implementations?
```suggestion
You can choose which implementation is used in `services.yml` file in your project.
```
